packages/vizbuilder/src/index.js

```diff
@@ -0,0 +1,112 @@
+import classnames from "classnames";
+import queryString from "query-string";
+
+import {fetchCubes} from "./actions/fetch";
+import {loadControl, setStatePromise} from "./actions/loadstate";
+import AreaChart from "./components/AreaChart";
+import AreaLoading from "./components/AreaLoading";
+import AreaSidebar from "./components/AreaSidebar";
+import {UIToaster} from "./components/UIToaster";
+import * as api from "./helpers/api";
+import initialState from "./state";
+
+import "@blueprintjs/labs/dist/blueprint-labs.css";
+import "./index.css";
+
+class Vizbuilder extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    api.resetClient(props.src);
+    this.state = initialState();
+
+    this.loadControl = loadControl.bind(this);
+    this.firstLoad = this.firstLoad.bind(this);
+    this.stateUpdate = this.stateUpdate.bind(this);
+  }
+
+  getChildContext() {
+    return {
+      loadControl: this.loadControl,
+      stateUpdate: this.stateUpdate
+    };
+  }
+
+  componentDidMount() {
+    const locationQuery = queryString.parse(location.search);
+    this.firstLoad(locationQuery);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const {src, datasetDidChange} = this.props;
+    const {dataset, load} = this.state;
+    const {error, severity} = load;
+
+    if (src && prevProps.src !== src) {
+      api.resetClient(src);
+      this.setState(initialState(), this.firstLoad);
+    }
+
+    if (datasetDidChange && dataset !== prevState.dataset) {
+      datasetDidChange(dataset);
+    }
+
+    if (error && prevState.load.error !== error) {
+      console.warn(error.stack);
+      UIToaster.show({intent: severity, message: error.message});
+    }
+  }
+
+  render() {
+    const {dataset, load, members, options, query, queryOptions} = this.state;
+    return (
+      <div className={classnames("vizbuilder", {loading: load.inProgress})}>
+        <AreaLoading {...load} />
+        <AreaSidebar options={options} query={query} queryOptions={queryOptions} />
+        <AreaChart dataset={dataset} members={members} query={query} />
+      </div>
+    );
+  }
+
+  stateUpdate(newState) {
+    return setStatePromise.call(this, state => {
+      const finalState = {};
+      const keys = Object.keys(newState);
+
+      for (let i = 0; i < keys.length; i++) {
+        const key = keys[i];
+        finalState[key] = {
+          ...state[key],
+          ...newState[key]
+        };
+      }
+
+      return finalState;
+    });
+  }
+
+  firstLoad(locationQuery) {
+    this.loadControl(fetchCubes.bind(this, locationQuery), () => {
+      const {query, queryOptions} = this.state;
+      return api.query({
+        ...query,
+        options: queryOptions
+      });
+    });
+  }
+}
+
+Vizbuilder.childContextTypes = {
+  loadControl: PropTypes.func,
+  stateUpdate: PropTypes.func
+};
+
+Vizbuilder.propTypes = {
+  src: PropTypes.string,
+  datasetWillChange: PropTypes.func,
+  datasetDidChange: PropTypes.func
```

Comment:
Those are lifecycle functions I'm planning to add.
When working on datachile, we sometimes had to hack our way into some modules to get certain information. Having lifecycle hooks available is always a good idea, even more in a capsuled component like vizbuilder, since there's no other way to get the internal state.
For example, if I need to access the current dataset from the parent app to keep a history, or to hook the current state and update the permalink.